### PR TITLE
ARROW-8169: [Java] Improve the performance of JDBC adapter by allocating memory proactively

### PR DESCRIPTION
**Pull request by liyafan82**
The current implementation use setSafe methods to dynamically allocate memory if necessary. For fixed width vectors (which are frequently used in JDBC), however, we can allocate memory proactively, since the vector size is known as a configuration parameter. So for fixed width vectors, we can use set methods instead.

This change leads to two benefits:
1. When processing each value, we no longer have to check vector capacity and reallocate memroy if needed. This leads to better performance.
2. If we allow the memory to expand automatically (each time by 2x), the amount of memory usually ends up being more than necessary. By allocating memory by the configuration parameter, we allocate no more, or no less.

Benchmark results show notable performance improvements:

Before:

Benchmark Mode Cnt Score Error Units
JdbcAdapterBenchmarks.consumeBenchmark avgt 5 521.700 ± 4.837 us/op

After:

Benchmark Mode Cnt Score Error Units
JdbcAdapterBenchmarks.consumeBenchmark avgt 5 430.523 ± 9.932 us/op

### CODE REVIEW

**Comment by tianchen92** [comment]
Nice catch, if this works, I should do similar work in Avro adapter(I created https://issues.apache.org/jira/browse/ARROW-8171 to track).
In the initial iterator design, one could convert data into multi Arrow batch or only one batch controled by the variable targetBatchSize（-1 or not), for cases targetBatchSize > 0, preallocate is better, however, when targetBatchSize == -1, the actual row count is not known.
In this PR, seems that converting into only one batch was discarded by the change in JdbcToArrowConfigBuilder, so I am not sure if it is reasonable

**Comment by liyafan82** [comment]
> Nice catch, if this works, I should do similar work in Avro adapter(I created https://issues.apache.org/jira/browse/ARROW-8171 to track).
> In the initial iterator design, one could convert data into multi Arrow batch or only one batch controled by the variable targetBatchSize（-1 or not), for cases targetBatchSize > 0, preallocate is better, however, when targetBatchSize == -1, the actual row count is not known.
> In this PR, seems that converting into only one batch was discarded by the change in JdbcToArrowConfigBuilder, so I am not sure if it is reasonable

Sure. Thanks for your kind reminder. We need to take care of the special case when targetBatchSize equals -1.

**Comment by emkornfield** [comment]
@liyafan82 This will need to be rebased.

**Comment by liyafan82** [comment]
> @liyafan82 This will need to be rebased.

Thanks for your kind reminder. 

> @liyafan82 This will need to be rebased.

Rebased. Thank you.

File: java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/DateConsumer.java [hunk 1]
```diff
@@ -89,7 +89,7 @@ public void consume(ResultSet resultSet) throws SQLException {
         if (day < 0 || day > MAX_DAY) {
           throw new IllegalArgumentException("Day overflow: " + day);
         }
-        vector.setSafe(currentIndex, day);
+        vector.set(currentIndex, day);
```
Comment (emkornfield) [inline java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/DateConsumer.java:92]: please add comments for each of the set methods to indicate why setSafe is not required.

Comment (liyafan82) [reply java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/DateConsumer.java:94]: Comments added. Thank you for the good suggestion.

**Review by rymurr — APPROVED** [review]

**Pull request closed without merging**